### ISSUE TITLE
Architecture. A proposal.

### ISSUE THREAD

**Issue by DavidBM** [Couragium/rust-anonymize-mysqldump#1]
Hi! 

I was thinking a bit about this projects. I guess the main point of this program will be to be able to pipe it. For the one written in go it is "easy" to do as it takes the position of the item to replace. This supposes several problems as, for example, when adding a new column to the DB. It would invalidate the anonymizer configuration.

That is why we spoke for this project to use column names instead of positions. In order to do that we need to parse the dump's table definitions and use that information to identify what field we should change.

In order to be able to pipe it, we need to work in a stream like way. That means that we may not have the whole backup data loaded when we are replacing some field. This may bring some problems as knowing what field to replace may not be evident (INSERT statements don't contain the column name).

My proposal is to build a very simple and short parser using [nom](https://crates.io/crates/nom) to split the table dumps and then, for each one, to split the table definition from the INSERT statements. 

Once done, I see three options:

- A) We can parse each (table definition and INSERT statements) using [sqlparser-rs](https://github.com/andygrove/sqlparser-rs), make the changes in the AST and then rebuild the query string. 
- B) We write another nom parser that allows us to parse the parts we need from an INSERT string and write a function what builds an almost identical INSERT statement with the values changed. 
- C) Decide what we are cooler and just use regexp. 

In my opinion the second ones is the faster and easier one, plays very well with Streams.

For the config files I would use something simple like this yaml:

```yaml
tableA.column_name: typeX
tableB.column_name: typeY
tableC.column_name:
  - typeZ
  - typeZ_configA_if_needed: value
    typeZ_configB_if_needed: value
```

What do you think?

**Comment by DavidBM** [comment]
I was playing with the code a bit. 

Reading stdin and copy to stdout: Check. Easy.
Parsin and getting the AST from an SQl statement: Check. Easy.

Parsing an stream because the file can be huge and getting the AST: Ups, difficult to do.
Parsing an stream because the file can be huge and getting the AST and having an API that allows to mutate the query during the stream processing/parsing of the file?: Bunkers. We will need to do our own parser :(

I'm kind of lost here to be honest. 

For our use case the best would be some kind of SQL parser where for each token/entity there is a callback called with context enough to decide what to do and that allows to mutate the token/entity in the same callback. I cannot find anything like that. I'm afraid we may need to build our own.

**Comment by DavidBM** [comment]
Well, come news. I have taken the SqlParser-rs crate and heavily modified it in order to parse using an stream of bytes (transformed to tokens as it advances) and I did some magic in order to allow this stream of tokens to be edited in real time with some context of what that token represents. 

This basically, allows to use the library to edit a mysqldump file at the same time it reads it.

Here the results: https://github.com/Couragium/mysqldump-mutator-rs

**Comment by DavidBM** [comment]
Good news:
I think https://github.com/Couragium/mysqldump-mutator-rs is finish as it is able to parse and serialize the Omnea Backup (4.5Gb) producing a perfect byte-by-byte copy of the original file in 4 minutes using 100Mb of ram. This was an important test as MySQL is a crazy SQL dialect and our database somehow uses it all (for example, we have a table that uses CSV engine instead of InnoDB) and being able to parse a mysqldump with all of the 4.5Gb of crazy stuff is not easy. The original parser would have used +6Gb of RAM and not being able to understand all the SQL. 

In short, we can start with this project as the hardcore complexity of understanding and mutating an mysqldump is solved. 

I'm closing this issue as it is solved with that library.
